File: Emulator/src/main/core/game/interaction/Option.java
package core.game.interaction;

import core.cache.def.impl.ItemDefinition;
import core.cache.def.impl.NPCDefinition;
import core.cache.def.impl.SceneryDefinition;
import core.game.node.Node;
import core.game.node.entity.npc.NPC;
import core.game.node.item.Item;
import core.game.node.scenery.Scenery;

/**
 * Represents a option on a node.
 */
public final class Option {

    /**
     * Player attack option.
     */
    public static final Option _P_ATTACK = new Option("Attack", 0);

    /**
     * Player follow option.
     */
    public static final Option _P_FOLLOW = new Option("Follow", 2);

    /**
     * Player trade option.
     */
    public static final Option _P_TRADE = new Option("Trade with", 3);

    /**
     * Player give-to option.
     */
    public static final Option _P_GIVETO = new Option("Give-to", 3);

    /**
     * Player pickpocket option.
     */
    public static final Option _P_PICKPOCKET = new Option("Pickpocket", 4);

    /**
     * Player examine option.
     */
    public static final Option _P_EXAMINE = new Option("Examine", 7);

    /**
     * Player assist request option.
     */
    public static final Option _P_ASSIST = new Option("Req Assist", 6);

    /**
     * Null fallback option.
     */
    public static final Option NULL = new Option("null", 0);

    private final String name;
    private final int index;
    private OptionHandler handler;

    /**
     * Creates a new Option with the given name and index.
     */
    public Option(String name, int index) {
        this.name = name;
        this.index = index;
    }

    /**
     * Gets the default handler for a node and option name.
     */
    public static OptionHandler defaultHandler(Node node, int nodeId, String name) {
        name = name.toLowerCase();
        if (node instanceof NPC) return NPCDefinition.getOptionHandler(nodeId, name);
        if (node instanceof Scenery) return SceneryDefinition.getOptionHandler(nodeId, name);
        if (node instanceof Item) return ItemDefinition.getOptionHandler(nodeId, name);
        return null;
    }

    /**
     * Gets the option name.
     */
    public String getName() {
        return name;
    }

    /**
     * Gets the option index.
     */
    public int getIndex() {
        return index;
    }

    /**
     * Gets the assigned handler.
     */
    public OptionHandler getHandler() {
        return handler;
    }

    /**
     * Sets and returns the handler.
     */
    public Option setHandler(OptionHandler handler) {
        this.handler = handler;
        return this;
    }
}

File: Emulator/src/main/core/game/interaction/Option.kt
package core.game.interaction

import core.cache.def.impl.ItemDefinition
import core.cache.def.impl.NPCDefinition
import core.cache.def.impl.SceneryDefinition
import core.game.node.Node
import core.game.node.entity.npc.NPC
import core.game.node.item.Item
import core.game.node.scenery.Scenery
import java.util.*

/**
 * Represents a option on a node.
 */
class Option(@JvmField val name: String, val index: Int) {
    /**
     * Gets the assigned handler.
     */
    var handler: OptionHandler? = null

    /**
     * Sets and returns the handler.
     */
    fun setHandler(handler: OptionHandler?): Option {
        this.handler = handler
        return this
    }

    companion object {
        /**
         * Player attack option.
         */
        val P_ATTACK = Option("Attack", 0)

        /**
         * Player follow option.
         */
        val P_FOLLOW = Option("Follow", 2)

        /**
         * Player trade option.
         */
        val P_TRADE = Option("Trade with", 3)

        /**
         * Player give-to option.
         */
        val P_GIVETO = Option("Give-to", 3)

        /**
         * Player pickpocket option.
         */
        val P_PICKPOCKET = Option("Pickpocket", 4)

        /**
         * Player examine option.
         */
        val P_EXAMINE = Option("Examine", 7)

        /**
         * Player assist request option.
         */
        val P_ASSIST = Option("Req Assist", 6)

        /**
         * Fallback/null option.
         */
        val NULL = Option("null", 0)

        /**
         * Gets the default handler for a node and option name.
         *
         * @param node The node the option applies to.
         * @param nodeId The definition ID of the node.
         * @param name The option name.
         * @return A matching option handler or null.
         */
        fun defaultHandler(node: Node?, nodeId: Int, name: String): OptionHandler? {
            val normalized = name.lowercase(Locale.getDefault())
            return when (node) {
                is NPC     -> NPCDefinition.getOptionHandler(nodeId, normalized)
                is Scenery -> SceneryDefinition.getOptionHandler(nodeId, normalized)
                is Item    -> ItemDefinition.getOptionHandler(nodeId, normalized)
                else       -> null
            }
        }
    }
}
